File: 2023/UICPC Regional/Round #14/solutions/H/kingMoritzFord.java
// @EXPECTED_RESULTS@: CORRECT
import java.util.*;

public class kingMoritzFord{
	static class Arc{
		public int cap;
		public int target;
	}
	static class pii{
		public int first;
		public int second;
		public pii( int _first, int _second ) { first = _first;  second = _second; }
	}
	public static boolean visited[];
	public static int parent[];
	public static int count[];
	public static Arc graph[][];
	public static int rows, cols, cas_y, cas_x;
	public static final int off_x[] = {0,-1,0,1};
	public static final int off_y[] = {-1,0,1,0};	
	public static LinkedList<pii> que;

	public static int MAP( int y, int x, boolean z){ return ((y) * cols + (x) + ((z) ? rows * cols : 0 ));  }

	public static void pushNode( int node, int cap, int par ){
		visited[node] = true;
		parent[node] = par;
		que.add( new pii( node, cap ) );
	}

	public static int bfs(){
		que.clear();
		for( int i = 0; i < (2*rows*cols); ++i ) visited[i] = false;
		//init by pushing all outer vertices
		for( int i = 1; i+1 < rows; ++i ){
			pushNode(MAP(i,0,false),20000000,-1);
			pushNode(MAP(i,cols-1,false),20000000,-1);
		}
		for( int i = 0; i < cols; ++i ){
			pushNode(MAP(0,i,false),20000000,-1);
			pushNode(MAP(rows-1,i,false),20000000,-1);
		}

		int target = MAP( cas_y, cas_x, true );
		while(que.size()!=0 && que.getFirst().first != target ){
			int cur = que.getFirst().first, min_cap = que.getFirst().second; que.remove();
			for( int i = 0; i < count[cur]; ++i ){
				Arc arc = graph[cur][i];
				if( !visited[arc.target] && (arc.cap > 0) ){
					int pc = (arc.cap < min_cap) ? arc.cap : min_cap;
					pushNode(arc.target, pc, cur );
				}
			}
		}
		if( que.size() != 0 ) //reached target
			return que.getFirst().second;
		return 0;
	}

	public static int getArc( int i, int j ){
		for( int k = 0; k < count[i]; ++k )
			if( graph[i][k].target == j )
				return k;
		return -1;
	}

	public static void reduce( int flow ){
		assert( flow > 0 );
		int node = MAP( cas_y, cas_x, true );
		while( parent[node] >= 0 ){
			//System.out.println( "Node: " + node + " Parent: " + parent[node] );
			int arc_id = getArc(parent[node],node), rev_arc_id = getArc(node,parent[node]);
			graph[parent[node]][arc_id].cap -= flow;
			graph[node][rev_arc_id].cap += flow;
			node = parent[node];
		}
	}

	public static void solve(){
		int flow = 0, min_cap;
		do{
			min_cap = bfs();
			if( min_cap > 0 ){
				flow += min_cap;
				reduce( min_cap );
			}
		} while( min_cap > 0 );
		System.out.println( flow );
	}

	public static void main( String args[] ){
		Scanner scanner = new Scanner( System.in );
		que = new LinkedList<pii>();
		rows = scanner.nextInt();
		cols = scanner.nextInt();
		count = new int[2 * rows * cols + 1 ];
		graph = new Arc[2 * rows * cols + 1 ][10];
		visited = new boolean[2 * rows * cols];
		parent = new int[ 2 * rows * cols ];
		Arc arc, infty_arc, zero_arc;
		for( int row = 0; row < rows; ++row ){
			for( int col = 0; col < cols; ++col ){
				int cap = scanner.nextInt(), node = MAP( row, col, false ), target = MAP( row, col, true );
				arc = new Arc();
				arc.target = target;
				arc.cap = cap;
				graph[node][count[node]++] = arc;
				arc = new Arc();
				arc.cap = 0;
				arc.target = node;
				graph[target][count[target]++] = arc;
			}
		}

		for( int row = 0; row < rows; ++row ){ for( int col = 0; col < cols; ++col ){
			for( int o = 0; o < 4; ++o ){
				if( (row + off_y[o] >= 0) && (row + off_y[o] < rows) && (col + off_x[o] >= 0) && (col+off_x[o] < cols) ){
					int out_node = MAP( row, col, true ), in_node = MAP( row + off_y[o], col + off_x[o], false );
					infty_arc = new Arc(); zero_arc = new Arc();
					infty_arc.cap = 20000000;
					zero_arc.cap = 0;
					zero_arc.target = out_node;
					infty_arc.target = in_node;
					graph[out_node][count[out_node]++] = infty_arc;
					graph[in_node][count[in_node]++] = zero_arc;
				}
			}
		} }
		cas_y = scanner.nextInt();
		cas_x = scanner.nextInt();
		solve();
	}
}

File: 2023/UICPC Regional/Round #14/solutions/H/king-moritz-ford.cpp
// @EXPECTED_RESULTS@: CORRECT
#include <iostream>
#include <climits>
#include <vector>
#include <algorithm>
#include <queue>
#include <utility>

using namespace std;

typedef pair<int,int> pii;
struct Arc{ int cap; int target; Arc() : cap(0), target(0){} };
vector< bool > visited;
vector< int > parent;
vector< vector<Arc> > graph;
int rows, cols, cas_y, cas_x;
pii off[] = {pii(-1,0),pii(0,-1),pii(1,0),pii(0,1)};

inline unsigned int MAP(int y,int x,bool z){ return ((y) * cols + (x) + ((z) ? rows * cols : 0 ));  }

void pushNode( int node, int cap, int par, queue<pii>& que ){
	visited[node] = true;
	parent[node] = par;
	que.push( pii( node, cap ) );
}

int bfs(){
	queue<pii> que;
	//init by pushing all outer vertices
	for( size_t i = 1; i+1 < rows; ++i ){
		pushNode(MAP(i,0,false),INT_MAX,-1,que);
		pushNode(MAP(i,cols-1,false),INT_MAX,-1,que);
	}
	for( size_t i = 0; i < cols; ++i ){
		pushNode(MAP(0,i,false),INT_MAX,-1,que);
		pushNode(MAP(rows-1,i,false),INT_MAX,-1,que);
	}

	int target = MAP( cas_y, cas_x, true );
	while(!que.empty() && que.front().first != target ){
		int cur = que.front().first, min_cap = que.front().second; que.pop();
		for( auto arc : graph[cur] ){
			if( !visited[arc.target] && arc.cap )
				pushNode(arc.target, std::min(arc.cap,min_cap), cur, que );
		}
	}
	for( size_t i = 0; i < visited.size(); ++i ) visited[i] = false;
	if( not que.empty() ){ //reached target
		return que.front().second;
	}
	return 0;
}

int getArc( int i, int j ){
	for( size_t k = 0; k < graph[i].size(); ++k )
		if( graph[i][k].target == j )
			return k;
	return -1;
}

void reduce( int flow ){
	int node = MAP( cas_y, cas_x, true );
	while( parent[node] >= 0 ){
		int arc_id = getArc(parent[node],node), rev_arc_id = getArc(node,parent[node]);
		graph[parent[node]][arc_id].cap -= flow;
		graph[node][rev_arc_id].cap += flow;
		node = parent[node];
	}
}

void solve(){
	int flow = 0, min_cap;
	do{
		min_cap = bfs();
		if( min_cap ){
			flow += min_cap;
			reduce( min_cap );
		}
	} while( min_cap );
	std::cout << flow << std::endl;
}

int main( int argc, char ** args ){
	while( cin >> rows >> cols ){
		graph.resize( 2 * rows * cols + 1 );
		visited.resize( 2 * rows * cols );
		parent.resize( 2 * rows * cols );
		Arc arc, infty_arc, zero_arc;
		for( int row = 0; row < rows; ++row ){
			for( int col = 0; col < cols; ++col ){
				cin >> arc.cap;
				int node = MAP( row, col, false ); arc.target = MAP( row, col, true );
				graph[node].push_back( arc );	//connect in and out node of (row,col)
				swap( arc.target, node );
				arc.cap = 0;
				graph[node].push_back(arc);
			}
		}
		infty_arc.cap = INT_MAX;
		zero_arc.cap = 0;
		for( int row = 0; row < rows; ++row ){ for( int col = 0; col < cols; ++col ){
			for( int o = 0; o < 4; ++o ){
				if( (row + off[o].first >= 0) && (row + off[o].first < rows) && (col + off[o].second >= 0) && (col+off[o].second < cols) ){
					int out_node = MAP( row, col, true ), in_node = MAP( row + off[o].first, col + off[o].second, false );
					zero_arc.target = out_node;
					infty_arc.target = in_node;
					graph[out_node].push_back( infty_arc );
					graph[in_node].push_back( zero_arc );
				}
			}
		} }
		cin >> cas_y >> cas_x;
		solve();
	}
	return 0;
}
